src/solver/expressions/visitors/CloneVisitor.cpp

```diff
@@ -28,9 +28,15 @@ CloneVisitor::CloneVisitor(Registry<Nodes::Node>& registry):
 {
 }
 
-Nodes::Node* CloneVisitor::visit(const Nodes::AddNode* node)
-{
-    return registry_.create<Nodes::AddNode>(dispatch(node->left()), dispatch(node->right()));
+Nodes::Node* CloneVisitor::visit(const Nodes::SumNode* node)
+{
+    std::vector<Nodes::Node*> clonedOperands;
+    clonedOperands.reserve(node->getSize());
+    for (auto* operand: node->getOperands())
+    {
+        clonedOperands.push_back(dispatch(operand));
+    }
+    return registry_.create<Nodes::SumNode>(clonedOperands);
```

Comment:
You may avoid a copy with a `std::move` here, if SumNode had move constructor. 

In `SumNode`
```cpp
SumNode(std::vector<Node*>&& operands) : operands_(std::move(operands)) {}
```
Here
```cpp
// Give ownership of clonedOperands to the caller
return registry_.create<Nodes::SumNode>(std::move(clonedOperands));
```